### PR TITLE
Edit DG for appendix b and c

### LINKED ISSUE #240: Move some planned enhancments to DG
Error handling: Improve error messages in edit command to check for existence of IC number in the event of empty fields.

Custom fields: Allow users to customize fields for patient records to suit their clinic's requirements.

Flexible Find command: Allow users to be able to find patients by their name, age, email, address, etc.

Updated Notes window: Update the Notes window to display the most updated note of the patient the user interacted with last.

Add Note features:
- Implement a feature to allow users to edit and delete specific notes.
- Allow users to know whose note they are viewing, by displaying the patient's name or IC number.
- Implement a feature for users to have the option of hiding the notes section in the patient's profile card.
- Allow users to tag and label different types of notes. For example, a tag for patient's diagnosis and another for their medication.

Optional fields: Allowing some fields e.g. email, phone number to be optional fields.

### PR DESCRIPTION
**Pull request by Jayne1010**
Closes #240 and #223 

- Merged optional fields with customisable fields 
- Added more support for language

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+In light of this issue, we would be working on our application to provide an update to the `find` command 
+to allow users to find contacts using different fields such as `name`, `age`, `email` etc. They will be case-insensitive to allow for better search results.
+
+**Examples**
+
+- find `John` returns contact with `name` of `John Doe` and corresponding `IC_NUMBER` of `T0123456A`
+- find `johndoe@mail.com` returns contact with `email` of `johndoe@mail.com` and corresponding `IC_NUMBER` of `T0123456A`
+
+### Update notes window during execution of `find` command
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: I think the update notes window is not referring to the `find` command because the `find` command does change the contents of out patient notes panel (the right hand side pane). 

This is more of after using find T0123456A then maybe u do addnote to S1234567A--> the patient notes panel will not show the updated note for S1234567A but will still show the note for T0123456A.

Then we wna say that we will want to fix this such that if we interact with any of the details of any patients (thru addnote/etc) then the most updated note will be displayed

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+In order to make it more convenient for the users, we will enhance `find` command such that on executing the `find` command, the notes of the first 
+person appearing at the top of the search results will replace the current notes in the notes panel. We will also allow users to know whose note they are viewing, by displaying the patient's name or IC number.
+
+**Examples**
+
+- If the notes panel is currently displaying the notes of the user with `IC_NUMBER` of `T1234567C`, 
+executing `find T0123456A`, will cause the notes panel to update to the notes of user with `IC_NUMBER` of `T0123456A`
+
+### Enabling optional fields
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: Maybe this one can merge with the first planned enhancement? because I feel like this is already mentioned a bit in the first enhancement.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+- `add n\John Doe p\12345678 i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25` will be allowed.
+- `add n\John Doe e\johndoe@mail.com i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25` will be allowed.
+- `add n\John Doe i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25` will be allowed.
+
+### Improving error handling for `edit` command
+
+**Potential Issues With Current Feature**
+
+Currently, the `edit` command does not check the existence of `IC_NUMBER` upon entering an `edit` command with an empty field.
+For instance, entering `edit S1234567P` in the event that the contact with the `IC_NUMBER` of `S1234567P` do not exist returns an error message
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: Can change this line to: For instance, entering `edit S1234567P` when the patient with the `IC_NUMBER` of `S1234567P` does not exist returns an error message...

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+- `add n\John Doe e\johndoe@mail.com i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25` will be allowed.
+- `add n\John Doe i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25` will be allowed.
+
+### Improving error handling for `edit` command
+
+**Potential Issues With Current Feature**
+
+Currently, the `edit` command does not check the existence of `IC_NUMBER` upon entering an `edit` command with an empty field.
+For instance, entering `edit S1234567P` in the event that the contact with the `IC_NUMBER` of `S1234567P` do not exist returns an error message 
+of *at least one field to edit must be provided*. However, the error message should be such that *the IC_NUMBER provided does
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: can remove the 'such'

Comment (tengcharmaine) [reply docs/DeveloperGuide.md]: Capitalise the at for 'At least one....' and the for 'The `IC_NUMBER` provided...' can

The IC_NUMBER also put in code block!

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+not exist* as the existence of the `IC_NUMBER` should be checked first. This is a problem as users might go on to provide a field for a non-existing contact, thus returning them another error message.
+
+**Proposed Enhancements**
+
+We will check the existence of the `IC_NUMBER` first during the execution of the `edit` command. 
+
+**Examples**
+
+`edit S1234567P` should return the error message *the IC_NUMBER provided does not exist.*
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: Can caps and add codeblock for the 'The `IC_NUMBER` provided...'

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+### Allowing for multiple phone numbers
+
+**Potential Issues With Current Feature**
+
+Currently, our `phone` number field only allows for one number. However, since out product is targeted at doctors working at the GP clinics with their contacts being patients, it
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: typo at the 'out' should be 'our'. But i think can change it to ClinicMate --> 'However, since ClinicMate is targeted at...'

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+**Potential Issues With Current Feature**
+
+Currently, our `phone` number field only allows for one number. However, since out product is targeted at doctors working at the GP clinics with their contacts being patients, it 
+is only natural for the `phone` number field to allow for multiple numbers. Reason being that a patient might have a caregiver or is required to contact his/her 
+family members. Therefore, restricting the `phone` to only one is very restrictive and could bring about inconvenience.
+
+**Proposed Enhancements**
+
+The team will thus allow for the addition of multiple numbers to be added under the `phone` field through the `add` command.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: Can change the 'The team' into 'We' and remove the 'thus' and 'to be added'

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+**Examples**
+
+- `add n\John Doe p\12345678 e\johndoe@mail.com i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25` will cause the `phone` number `12345678` to be added for user `T0123456A`
+- `add n\John Doe p\12345678, 09876544, 99999999 e\johndoe@mail.com i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25` will cause the `phone` number `12345678, 09876544, 99999999` to be added for user `T0123456A`
+
+### Allowing for derivation of age through Date-Of-Birth
+
+**Potential Issues With Current Feature**
+
+Existing features require the users to manually ass the age of contacts. For example, `add n\John Doe p\12345678 e\johndoe@mail.com i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25`.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: i think 'ass' should be 'enter'?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+However, it if very cumbersome for users to edit the age of their contact individually through the `edit` command.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: 'if' should be 'is'

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+**Proposed Enhancements**
+
+In order to allow to save the time of our users, we will adding in a new field called `DOB` which Date of Birth in the form of `DD/MM/YYYY`.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: add 'GPs' after the 'to allow' --> 'In order to allow GPs to...'

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+**Potential Issues With Current Feature**
+
+Existing features require the users to manually ass the age of contacts. For example, `add n\John Doe p\12345678 e\johndoe@mail.com i\T0123456A ag\12 s\M a\311, Clementi Ave 2, #02-25`.
+However, it if very cumbersome for users to edit the age of their contact individually through the `edit` command.
+
+**Proposed Enhancements**
+
+In order to allow to save the time of our users, we will adding in a new field called `DOB` which Date of Birth in the form of `DD/MM/YYYY`. 
+`age` of the contacts will thus be derived from their `DOB` with referenced to the current year which we are in. Hence, streamlining the contact management process.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: 'with reference'

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+Given that we are in the year 2024, the patient with the `DOB` of 
+
+- `10/10/2003` will be at the age of 21 
+- `12/12/1990` will be at the age of 34
+
+### Further enhancements for `addnote` command
+
+**Potential Issues With Current Feature**
+
+There were many constructive feedbacks given regarding the `addnote` command.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: I think can remove this line

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+Firstly, our current `addnote` command only allows user to add a new note, which will be appended to the previous one. However,
+there might be many instances in which the users need to edit or delete specific note. Disallowing such commands will make contact quite frustrating at times
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: can add a 'a' --> '...in which users need to edit or delete a specific note'

Comment (tengcharmaine) [reply docs/DeveloperGuide.md]: the last sentence can change to 'Disallowing this will make adding notes quite frustrating at times.'

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+Secondly, our patient profile card has a notes section which cannot be hidden and will appear in the UI. Users thus might not want to see the notes section in the patient's profile card as it might be very long and distracting, making it unpleasant to the eyes.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: Secondly, each of the contacts in the patient list panel has a notes section which cannot be hidden and will appear in the UI. Users might not want to see all the notes for each contact as it might be very long and distracting.

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -626,7 +626,161 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+Lastly, there is a lack of organisation for the notes and users might want to categorize their notes into different sections.
+
+**Proposed Enhancements**
+
+We will thus:
+- Implement a feature to allow users to edit and delete specific notes.
+- Implement a feature for users to have the option of hiding the notes section in the patient's profile card.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: Implement a feature for users to have the option of hiding the notes section for each patient in the patient list panel.

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -726,3 +880,60 @@ testers are expected to do more *exploratory* testing.
...
+### Effort 
+
+#### New Addition to classes
+New classes were added which enhanced ClinicMate's Features. Below are some examples of the new classes added:
+
+`Model` classes: IdentityCardNumber, Sex, Note 
+
+`Command` classes: show, addnote
+
+We ensured that these new classes follows the code quality standard strictly and practice defensive programming.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: follows --> follow; practice --> practiced

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -726,3 +880,60 @@ testers are expected to do more *exploratory* testing.
...
+A couple of test cases were also crafted in order to ensure that the new classes function as expected and meet the specified requirements. These test cases cover various scenarios and edge cases to validate the robustness and reliability of the code. 
+
+#### Enhancements to existing functions
+
+Other than making changes to the existing commands by adding in more parameters, we put in significant effort to make sure the edited commands works to fit the features of ClinicMate.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: edited --> existing?; can remove the 'works to' and change 'fit' to 'fits'

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -726,3 +880,60 @@ testers are expected to do more *exploratory* testing.
...
+For instance, the `find` command will allow users to find contact details with their IC, and more importantly, the user's notes will also appear on the notes panel on te right side of the split pane.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: 'and more importantly, the user's notes will also appear on the patient notes panel`

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -726,3 +880,60 @@ testers are expected to do more *exploratory* testing.
...
+This allowed for a more comprehensive view of contact's notes which is an important field which our targeted users will require. Hence, there were much effort placed into enhancing existing features to improve usability of ClinicMate.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: 'which is an important field that GPs will require.'

Comment (tengcharmaine) [reply docs/DeveloperGuide.md]: 'Hence, much effort was placed into enhancing existing features to improve usability of ClinicMate.'

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -726,3 +880,60 @@ testers are expected to do more *exploratory* testing.
...
+
+
+#### Improve User Interface of ClinicMate 
+
+Rather than the dark theme in which ClinicMate originally had, we chose to go with a lighter theme and a bigger font size in order to improve readability of ClinicMate.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: 'Rather than the dark theme which ClinicMate originally had,' --> removed the in

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -726,3 +880,60 @@ testers are expected to do more *exploratory* testing.
...
+Effort was put in by the team to learn different JavaFX functions such as the `ScrollPane` and `SplitPanel` in order to allow for a wider and better view of patient details in ClinicMate.
```
Comment (tengcharmaine) [inline docs/DeveloperGuide.md]: can 'the team' to 'us'

**Review by tengcharmaine — CHANGES_REQUESTED** [review]
Some part needs changes!

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (Jayne1010) [reply docs/DeveloperGuide.md]: I prefer the original one thoughh

Comment (tengcharmaine) [reply docs/DeveloperGuide.md]: okay can also keep ur original version then can change the 'do' to 'does' and the 'returns an error...' to 'returns the error...'

Comment (Jayne1010) [reply docs/DeveloperGuide.md]: ok yay will do thank youuu

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -626,7 +626,164 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+- find `John` returns contact with `name` of `John Doe` and corresponding `IC_NUMBER` of `T0123456A`
+- find `johndoe@mail.com` returns contact with `email` of `johndoe@mail.com` and corresponding `IC_NUMBER` of `T0123456A`
+
+### Update notes window during execution of `find` command
+
+**Potential Issues With Current Feature**
+
+Users have given us the feedback that our notes window on the right hand side of the split pane does not update with the use of other commands.
+For instance, if our notes panel currently already have `John`'s notes reflected on it, using the `addnote` command to add a note of another contact who is not `John` will not cause the 
+notes panel to update accordingly. `John`'s notes will still exist on the notes panel despite the executing the `addnote` command to edit the note of another person.
```
Comment (drustanyjt) [inline docs/DeveloperGuide.md]: "despite the executing the"

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -626,7 +626,164 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+**Examples**
+
+- find `John` returns contact with `name` of `John Doe` and corresponding `IC_NUMBER` of `T0123456A`
+- find `johndoe@mail.com` returns contact with `email` of `johndoe@mail.com` and corresponding `IC_NUMBER` of `T0123456A`
+
+### Update notes window during execution of `find` command
+
+**Potential Issues With Current Feature**
+
+Users have given us the feedback that our notes window on the right hand side of the split pane does not update with the use of other commands.
```
Comment (drustanyjt) [inline docs/DeveloperGuide.md]: notes window -> patient notes panel

**Review by drustanyjt — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.md [hunk 24]
```diff
@@ -726,3 +883,60 @@ testers are expected to do more *exploratory* testing.
...
+
+Other than making changes to the existing commands by adding in more parameters, we put in significant effort to make sure the edited commands works fits the features of ClinicMate.
+For instance, the `find` command will allow users to find contact details with their IC, and more importantly, and more importantly, the user's notes will also appear on the patient notes panel.
+This allowed for a more comprehensive view of contact's notes which is an important field which is an important field that GPs will require. Hence, much effort was placed into enhancing existing features to improve usability of ClinicMate.
+
+
+#### Improve User Interface of ClinicMate 
+
+Rather than the dark theme which ClinicMate originally had, we chose to go with a lighter theme and a bigger font size in order to improve readability of ClinicMate.
+Effort was put in by us to learn different JavaFX functions such as the `ScrollPane` and `SplitPanel` in order to allow for a wider and better view of patient details in ClinicMate.
```
Comment (drustanyjt) [inline docs/DeveloperGuide.md]: change functions to markup elements

**Review by drustanyjt — APPROVED** [review]
LGTM

**Review by tengcharmaine — APPROVED** [review]
LGTM

**Review by jiahui0309 — APPROVED** [review]
LGTM
